---
title: "Homework 2 - Bioconductor objects"
author: "your name here"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: html_document
---

# Question 1 - simple subsetting

Make an `ExpressionSet` called `eset` with the following tables. Make
a new version of the `eset` with just samples where treated is equal
to `"1"`, and the genes with IDs `"9"` and `"10"`. Show the `pData`,
`fData` and `exprs` tables for the subsetted object.

```{r}
exprs <- matrix(rnorm(6 * 10), ncol=6, nrow=10)
phenoData <- data.frame(sample=factor(1:6),
                        condition=factor(c("A","A","B","B","C","C")),
                        treated=factor(rep(0:1,3)))
featureData <- data.frame(geneID=1:10, geneSymbol=letters[1:10])
```

# Question 2 - gene density 

Start by loading the Ensembl database package we used in class. Pull
out the genes as a simple GRanges. Make a new GRanges `g2`, with the
genes on chromosome `"1"` (there are 5264 of these genes). Look up the
length of chromosome 1 using `seqinfo`. Make a `query` GRanges which
covers chromosome 1 with non-overlapping ranges of width 1 Mb (one
million letters). You should start with position 1, and end on
position 248,000,000 (instead of going to the very end of chromosome
1). Count the number of genes overlapping each range (you can use
`findOverlaps` and `table`) and then plot the counts (after converting
to a numeric vector). 

Do you see correlations in the counts along chromosome 1 or do the
counts look like independent draws?

What is the average number of genes per 1 Mb?

```{r}
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86
```

# Question 3 - exons per transcript

Construct the exons-by-transcript GRangesList object from the `edb`
database above. You can find the number of exons per transcript using
the `elementNROWS` function, which works on list-like objects. It
literally asks how many "rows" of data per element in the list, which
here is the number of exons per transcript. What is the mean and
median number of exons per transcript? Plot the histogram of the
number of exons per transcript. Sometimes `barplot(table(x))` is
better than a histogram. Run `barplot(table(...))` to plot the number
of exons per transcript, just from number of exons in the range
[1-15]. 

What is the maximum number of exons for a transcript? 

Pull out the exons (a GRanges object) for the transcript with the most
number of exons.

